web/src/js/ducks/modes/utils.ts

```diff
@@ -0,0 +1,81 @@
+import { getMode as getRegularModeConfig } from "./regular";
+import { getMode as getLocalModeConfig } from "./local";
+import { fetchApi } from "../../utils";
+
+export interface ModeState {
+    active: boolean;
+    name: string;
+    listen_port?: number;
+    listen_host?: string;
+    error?: string;
+}
+
+export const updateMode = () => {
+    return async (_, getState) => {
+        try {
+            const modes = getState().modes;
+
+            const activeModes: string[] = [
+                ...getRegularModeConfig(modes),
+                ...getLocalModeConfig(modes),
+                //add new modes here
+            ];
+            const response = await fetchApi.put("/options", {
+                mode: activeModes,
+            });
+            if (response.status === 200) {
+                return { success: true };
+            } else {
+                const errorText = await response.text();
+                return { success: false, error: errorText };
+            }
+        } catch (error) {
+            //TODO: handle error
+        }
+    };
+};
+
+export const addListenAddr = (mode: ModeState) => {
+    let stringMode = mode.name;
+    if (mode.active) {
+        if (mode.listen_host && mode.listen_port) {
+            stringMode += `@${mode.listen_host}:${mode.listen_port}`;
+        } else if (mode.listen_port) {
+            stringMode += `@${mode.listen_port}`;
+        }
+        return [stringMode];
+    }
+    return [];
+};
+
+const parseMode = (modeConfig) => {
```

Comment:
parseMode needs dedicated tests and probably needs to follow the logic in https://github.com/mitmproxy/mitmproxy/blob/v10.3.1/mitmproxy/proxy/mode_specs.py#L102-L123 more closely.

See https://github.com/mitmproxy/mitmproxy/blob/main/test/mitmproxy/proxy/test_mode_specs.py for inspiration. :) It's a bit easier for us because we can expect well-formed data. :)